pkg/northbound/gnmi/subscribe.go

```diff
@@ -44,79 +52,105 @@ func (s *Server) Subscribe(stream gnmi.GNMI_SubscribeServer) error {
+				//Ignoring Errors during removal
+				mgr.Dispatcher.UnregisterNbi(hash)
+				mgr.Dispatcher.UnregisterOperationalState(hash)
+				resChan <- result{success: true, err: nil}
+				break
+			}
 
-		if err != nil {
-			log.Println("Error in subscription", err)
-			//Ignoring Errors during removal
-			mgr.Dispatcher.UnregisterNbi(hash)
-			mgr.Dispatcher.UnregisterOperationalState(hash)
-			return err
-		}
+			if err != nil {
+				code, ok := status.FromError(err)
+				if ok && code.Code() == codes.Canceled {
+					log.Println("Subscription Terminated")
+					resChan <- result{success: true, err: nil}
+				} else {
+					log.Println("Error in subscription", err)
+					//Ignoring Errors during removal
+					mgr.Dispatcher.UnregisterNbi(hash)
+					mgr.Dispatcher.UnregisterOperationalState(hash)
+					resChan <- result{success: false, err: err}
+				}
+				break
+			}
 
-		var mode gnmi.SubscriptionList_Mode
+			var mode gnmi.SubscriptionList_Mode
 
-		if in.GetPoll() != nil {
-			mode = gnmi.SubscriptionList_POLL
-		} else {
-			subscribe = in.GetSubscribe()
-			mode = subscribe.Mode
-		}
-		//If the subscription mode is ONCE or POLL we immediately start a routine to collect the data
-		if mode != gnmi.SubscriptionList_STREAM {
-			go collector(stream, subscribe)
-		} else {
-			subs := subscribe.Subscription
-			//FAST way to identify if target and subscription is present
-			subsStr := make(map[string]struct{})
-			targets := make(map[string]struct{})
-			for _, sub := range subs {
-				subscriptionPathStr := utils.StrPath(sub.Path)
-				subsStr[subscriptionPathStr] = struct{}{}
-				targets[sub.Path.Target] = struct{}{}
+			if in.GetPoll() != nil {
+				mode = gnmi.SubscriptionList_POLL
+			} else {
+				subscribe = in.GetSubscribe()
+				mode = subscribe.Mode
+			}
+			//If the subscription mode is ONCE or POLL we immediately start a routine to collect the data
+			if mode != gnmi.SubscriptionList_STREAM {
+				go collector(stream, subscribe, resChan)
+			} else {
+				subs := subscribe.Subscription
+				//FAST way to identify if target and subscription is present
+				subsStr := make(map[string]struct{})
+				targets := make(map[string]struct{})
+				for _, sub := range subs {
+					subscriptionPathStr := utils.StrPath(sub.Path)
+					subsStr[subscriptionPathStr] = struct{}{}
+					targets[sub.Path.Target] = struct{}{}
+				}
+				//Each subscription request spawns a go routing listening for related events for the target and the paths
+				go listenForUpdates(changesChan, stream, mgr, targets, subsStr, resChan)
+				go listenForOpStateUpdates(opStateChan, stream, targets, subsStr, resChan)
 			}
-			//Each subscription request spawns a go routing listening for related events for the target and the paths
-			go listenForUpdates(changesChan, stream, mgr, targets, subsStr)
-			go listenForOpStateUpdates(opStateChan, stream, targets, subsStr)
 		}
-	}
+	}()
 
+	for result := range resChan {
+		if !result.success {
```

Comment:
will this wait for its child threads to **join** before it can complete?

Reply:
there is just one go routine, synchronisation is handled but the "result" message. There will be one result message out of the resChan and will be handled here.